### ISSUE TITLE
Add support for regression problems

### ISSUE THREAD

**Issue by weixuanfu** [EpistasisLab/pennai#229]
- [x] Dataset upload changes (Heather)
- [x] Knowledge base of regression benchmarks (Bill)
- [x] Result pages of regression problem (Weixuan and Josh)
- [x] Experiments page change for regression algorithms  (Heather)
- [x] API of dataset with a new terms of problem types ("classification" or "regression")(Weixuan and Heather)
- [x] Add regression unit tests for Recommender Engines (Bill)
- [x] Update recommender engine to check that the dataset passed to an update() or recommend() method matches the expected prediction_type (Bill)
- [x] Update AI engine to create regression and classification recommender engine instances.  Direct datasets for update() and recommend() requests by prediction_type to go to the correct instance (Heather and Bill)

**Comment by weixuanfu** [comment]
https://github.com/EpistasisLab/tpot/blob/master/tpot/config/regressor.py

**Comment by weixuanfu** [comment]
@lacava I added 5 more regessors to projects.json in this commit ceab1ceccd038248582db7ca23adc9ee406e3ae7. Please check whether the parameters are proper.

@hjwilli Thank you for updating API of dataset. I just merged the regresion_229 branch to my working branch for further tests.

**Comment by lacava** [comment]
some comments:

- [100, 250] for the ensemble trees seems kinda low. how bout [100, 500]?
- min_weight_fraction_leaf probably doesn't need to go up to 0.5
- SVC: make "C" a log scale, maybe np.logspace(-4,2,5)
- same with gamma
- i'm not sure coef0 needs to go to 100 or have a 50 in there (make log scale)
- might want to set max_iters in LassoLarsCV

**Comment by hjwilli** [comment]
Changes in branch [regression_229](https://github.com/EpistasisLab/pennai/tree/regression_229):
- `prediction_type` (classification or regression) for a dataset is now stored serverside.  POST/GET calls to `/datasets/*` have been updated appropriately, see api.raml for reference.
- dataset upload validation updated for regression problems
- initial upload of datasets from the `data/datasets/user` directory now supports regression problems, see docs for dataset config specification

---
Hi @weixuanfu, please merge the latest changes from regression_229 to your working branch.

**Comment by weixuanfu** [comment]
@hjwilli OK, actually I merged the branch to my working branch reg_api before (see the comments above) and it works. I will update my working branch with latest changes from regression_229.

**Comment by weixuanfu** [comment]
> 
> 
> some comments:
> 
>     * [100, 250] for the ensemble trees seems kinda low. how bout [100, 500]?
I kept `[100, 250]` for matching this parameters for classification problems. 
> 
>     * min_weight_fraction_leaf probably doesn't need to go up to 0.5

I removed 0.5 for all `min_weight_fraction_leaf ` parameters 
> 
>     * SVC: make "C" a log scale, maybe np.logspace(-4,2,5)
> 

Updated to [0.0001, 0.001, 0.01, 0.1, 1, 10, 100]
>     * same with gamma
> 
Updated to [0.0001, 0.001, 0.01, 0.1, 1, 10, 100]
>     * i'm not sure coef0 needs to go to 100 or have a 50 in there (make log scale)
> 
Updated to [0.0, 0.0001, 0.001, 0.01, 0.1, 1, 10]
>     * might want to set max_iters in LassoLarsCV
Already set to 10000

**Comment by lacava** [comment]
can we add kernel ridge regression with a guassian kernel?

**Comment by weixuanfu** [comment]
> can we add kernel ridge regression with a guassian kernel?

OK, looks good.

**Comment by weixuanfu** [comment]
> 
> 
> can we add kernel ridge regression with a guassian kernel?

@lacava it is a little tricky to add Gaussian kernel (check [this example](https://scikit-learn.org/stable/auto_examples/gaussian_process/plot_compare_gpr_krr.html#sphx-glr-auto-examples-gaussian-process-plot-compare-gpr-krr-py)) in [`KernelRidge`](https://scikit-learn.org/stable/modules/generated/sklearn.kernel_ridge.KernelRidge.html#sklearn.kernel_ridge.KernelRidge) because it seems only supports "linear" kernel as string input but support other kernels by using callable objects instead. But PennAI do not support callable inputs defined in projects.json so far.

**Comment by lacava** [comment]
@weixuanfu I don't think that's correct. this works fine (on 0.21.3)

```
from sklearn.kernel_ridge import KernelRidge
ml = KernelRidge(kernel='rbf', alpha=0.1, gamma=0.01)
from sklearn.datasets import load_boston
X,y = load_boston(return_X_y=True)
ml.fit(X,y)
ml.predict(X)
```

I would suggest these parameters:

    hyper_params = {
            'kernel': ['rbf'],
            "alpha": [1e0, 0.1, 1e-2, 1e-3],
            "gamma": np.logspace(-2, 2, 5)
            }

**Comment by weixuanfu** [comment]
- [ ] add pearson r squared for regression
- [ ] default scores for regression is MSE

**Comment by lacava** [comment]
@weixuanfu the parameters for XGBoost are not correct in projects.json 

these parameters are not valid:
['min_samples_split', 'min_samples_leaf', 'max_features']

available params for XGBoostRegressor:
```python
['base_score', 'booster', 'colsample_bylevel', 'colsample_bynode', 'colsample_bytree', 'gamma', 'importance_type', 'learning_rate', 'max_delta_step', 'max_depth', 'min_child_weight', 'missing', 'n_estimators', 'n_jobs', 'nthread', 'objective', 'random_state', 'reg_alpha', 'reg_lambda', 'scale_pos_weight', 'seed', 'silent', 'subsample', 'verbosity']
```

**Comment by weixuanfu** [comment]
@lacava Thank you for catching this bug. I fixed the issue via the commit dc459e0f5c0b3220d22cc764da23f6180db55b47 above

**Comment by lacava** [comment]
it's checked, but the experiments page is not displaying the accuracy of the regression experiments. 
![image](https://user-images.githubusercontent.com/6578458/70570260-8bd92000-1b69-11ea-94ed-f3ea502c8648.png)

**Comment by lacava** [comment]
@weixuanfu after looking at the results page on my computer, i think it would be better to have the QQ plot in a separate block. It is actually shorter than the other columns on my desktop, but of course it scales with the aspect ratio of your browser.

**Comment by weixuanfu** [comment]
> it's checked, but the experiments page is not displaying the accuracy of the regression experiments.
> ![image](https://user-images.githubusercontent.com/6578458/70570260-8bd92000-1b69-11ea-94ed-f3ea502c8648.png)
@lacava , we need redesign the table in experiments page since there is no accuracy for regression. What is default score metrics for regression problems in AI?

**Comment by lacava** [comment]
@weixuanfu i have the default set to r2_cv_mean.

**Comment by weixuanfu** [comment]
@lacava AI were not working after merging this commit f387b1c. Below was the stderr I caught. Any idea?

```shell
1|ai     | ai: INFO: 2019 02:30:18 PM UTC: 1 new results!
1|ai     | api_utils: INFO: get_metafeatures(5df24dcf2143930042e3c234)
0|lab    | GET /api/datasets/5df24dcf2143930042e3c234 200 - - 2.617 ms
1|ai     | api_utils:get_metafeatures
1|ai     | ai: INFO: mf count:
1|ai     |  1
1|ai     | ai: INFO: indices: 
1|ai     |  ['5df24dcf2143930042e3c234']
1|ai     | surprise_recommenders: INFO: updating SVD model
1|ai     | ai: ERROR: Unhanded exception caught: <class 'KeyError'>
1|ai     | ai: INFO: Shutting down AI engine...
1|ai     | ai: INFO: ...Shutting down Request Manager...
1|ai     | ai: INFO: Goodbye
1|ai     | Traceback (most recent call last):
1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
1|ai     |     "__main__", mod_spec)
1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
1|ai     |     exec(code, run_globals)
1|ai     |   File "/appsrc/ai/ai.py", line 595, in <module>
1|ai     |     main()
1|ai     |   File "/appsrc/ai/ai.py", line 571, in main
1|ai     |     pennai.update_recommender()
1|ai     |   File "/appsrc/ai/ai.py", line 328, in update_recommender
1|ai     |     self.rec_engines[predictionType].update(filterRes, filterMf)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 81, in update
1|ai     |     self.update_model(results_data)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 283, in update_model
1|ai     |     self.update_training_data(results_data)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 91, in update_training_data
1|ai     |     results_data[['algorithm-parameters','_id','score']]
1|ai     |   File "/usr/local/lib/python3.7/site-packages/pandas/core/frame.py", line 2934, in __getitem__
1|ai     |     raise_missing=True)
1|ai     |   File "/usr/local/lib/python3.7/site-packages/pandas/core/indexing.py", line 1354, in _convert_to_indexer
1|ai     |     return self._get_listlike_indexer(obj, axis, **kwargs)[1]
1|ai     |   File "/usr/local/lib/python3.7/site-packages/pandas/core/indexing.py", line 1161, in _get_listlike_indexer
1|ai     |     raise_missing=raise_missing)
1|ai     |   File "/usr/local/lib/python3.7/site-packages/pandas/core/indexing.py", line 1252, in _validate_read_indexer
1|ai     |     raise KeyError("{} not in index".format(not_found))
1|ai     | KeyError: "['score'] not in index"
```

**Comment by weixuanfu** [comment]
![image](https://user-images.githubusercontent.com/21084970/70808417-a5f34800-1d8d-11ea-8291-1d80fff89a0d.png)
I try to update the experiment table as above. Please let me know if you have any idea.

**Comment by lacava** [comment]
@weixuanfu I just renamed the 'accuracy' frame to 'r2_cv_mean' to match the knowledgebase. that might affect your code.

**Comment by lacava** [comment]
@weixuanfu I think it would make sense to change the column header to 'Score' with an info box that shows what the score is for regression and classification.

**Comment by weixuanfu** [comment]
> 
> 
> @weixuanfu I just renamed the 'accuracy' frame to 'r2_cv_mean' to match the knowledgebase. that might affect your code.

I saw that commit and thank you for the quick fix.

I also saw that regression knowledge base is much larger than classification one (see the log below), which make it very slow to load and it may be the reason that it failed in the integration tests. Could we fix that?
```shell
1|ai     | knowledgebase_loader: INFO: returning 79608 results from data/knowledgebases/sklearn-benchmark5-data-knowledgebase.tsv.gz
1|ai     | knowledgebase_loader: INFO: returning 1467536 results from data/knowledgebases/pmlb_regression_results.tsv.gz```

**Comment by weixuanfu** [comment]
```shell
1|ai     | ai: INFO: Goodbye
1|ai     | Traceback (most recent call last):
1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
1|ai     |     "__main__", mod_spec)
1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
1|ai     |     exec(code, run_globals)
1|ai     |   File "/appsrc/ai/ai.py", line 595, in <module>
1|ai     |     main()
1|ai     |   File "/appsrc/ai/ai.py", line 571, in main
1|ai     |     pennai.update_recommender()
1|ai     |   File "/appsrc/ai/ai.py", line 328, in update_recommender
1|ai     |     self.rec_engines[predictionType].update(filterRes, filterMf)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 81, in update
1|ai     |     self.update_model(results_data)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 283, in update_model
1|ai     |     self.update_training_data(results_data)
1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 97, in update_training_data
1|ai     |     self.reader, rating_scale=(0,1))
1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 193, in load_from_df
1|ai     |     return DatasetAutoFolds(reader=reader, df=df, rating_scale=rating_scale)
1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 317, in __init__
1|ai     |     self.df.itertuples(index=False)]
1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 316, in <listcomp>
1|ai     |     for (uid, iid, r) in
1|ai     | TypeError: float() argument must be a string or a number, not 'tuple'
```
@lacava AI still failed somehow.

**Comment by hjwilli** [comment]
Hi @lacava, I'm also seeing the error that Weixuan references above.  I've been able to reproduce by starting the app with the svd recommender and either running a manual experiment or turning on the ai for a classification dataset.  I haven't seen the error with a regression dataset.  It looks like the error occurs when `pennai.update_recommender()` is called for a classification result on one of the surprise recommenders.  Example error log:

```
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe 200 15 - 3.763 ms
lab_1      | 0|lab    | GET /api/userexperiments/5df9468936a9990035cb1cfe 200 2223 - 23.092 ms
lab_1      | 0|lab    | GET /api/userdatasets/5df3bc234aa282003bdebaa1 200 2166 - 17.553 ms
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/imp_score_5df9468936a9990035cb1cfe.png
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe/files 200 28 - 4.542 ms
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/dtree_5df9468936a9990035cb1cfe.dot
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe/files 200 28 - 3.841 ms
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/dtree_5df9468936a9990035cb1cfe.png
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/cnf_matrix.json
lab_1      | 0|lab    | serverSocket.emitEvent('expUpdated', '[object Object]')
lab_1      | 0|lab    | { collection: 'experiments', id: '5df9468936a9990035cb1cfe' }
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe 200 15 - 4.054 ms
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe/files 200 28 - 13.629 ms
lab_1      | 0|lab    | GET /api/userexperiments/5df9468936a9990035cb1cfe 200 2827 - 14.526 ms
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/confusion_matrix_5df9468936a9990035cb1cfe.png
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe/files 200 28 - 2.977 ms
lab_1      | 0|lab    | GET /api/userdatasets/5df3bc234aa282003bdebaa1 200 2166 - 13.907 ms
lab_1      | 1|ai     | ai: INFO: 2019 09:20:12 PM UTC: checking results...
lab_1      | 1|ai     | api_utils: INFO: get_new_experiments_as_dataframe(1576617575000)
lab_1      | 1|ai     | api_utils: INFO: get_new_experiments(1576617575000)
lab_1      | 0|lab    | POST /api/experiments 200 - - 1.703 ms
lab_1      | 1|ai     | ai: INFO: 2019 09:20:12 PM UTC: 1 new results!
lab_1      | 1|ai     | ai: INFO: mf count:
lab_1      | 1|ai     |  2
lab_1      | 1|ai     | ai: INFO: indices:
lab_1      | 1|ai     |  ['5df3bc234aa282003bdebaa1']
lab_1      | 1|ai     | surprise_recommenders: INFO: updating SVD model
lab_1      | 1|ai     | ai: ERROR: Unhanded exception caught: <class 'TypeError'>
lab_1      | 1|ai     | ai: INFO: Shutting down AI engine...
lab_1      | 1|ai     | ai: INFO: ...Shutting down Request Manager...
lab_1      | 1|ai     | request_manager: INFO: AiRequest terminate_request (appendicitis,5df3bc234aa282003bdebaa1)
lab_1      | 1|ai     | api_utils: INFO: set_ai_status(5df3bc234aa282003bdebaa1, finished)
lab_1      | 0|lab    | serverSocket.emitEvent('aiToggled', '[object Object]')
lab_1      | 0|lab    | { id: '5df3bc234aa282003bdebaa1' }
lab_1      | 0|lab    | =socketServer:aiToggled(5df3bc234aa282003bdebaa1)
lab_1      | 0|lab    | PUT /api/userdatasets/5df3bc234aa282003bdebaa1/ai 200 53 - 4.444 ms
lab_1      | 1|ai     | ai: INFO: Goodbye
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/roc_curve_5df9468936a9990035cb1cfe.png
lab_1      | 1|ai     | Traceback (most recent call last):
lab_1      | 1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 193, in _run_module_as_main
lab_1      | 1|ai     |     "__main__", mod_spec)
lab_1      | 1|ai     |   File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
lab_1      | 1|ai     |     exec(code, run_globals)
lab_1      | 1|ai     |   File "/appsrc/ai/ai.py", line 595, in <module>
lab_1      | 1|ai     |     main()
lab_1      | 1|ai     |   File "/appsrc/ai/ai.py", line 571, in main
lab_1      | 1|ai     |     pennai.update_recommender()
lab_1      | 1|ai     |   File "/appsrc/ai/ai.py", line 328, in update_recommender
lab_1      | 1|ai     |     self.rec_engines[predictionType].update(filterRes, filterMf)
lab_1      | 1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 81, in update
lab_1      | 0|lab    | PUT /api/v1/experiments/5df9468936a9990035cb1cfe/files 200 28 - 11.885 ms
lab_1      | 1|ai     |
lab_1      | 1|ai     | self.update_model(results_data)
lab_1      | 1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 283, in update_model
lab_1      | 1|ai     |
lab_1      | 1|ai     | self.update_training_data(results_data)
lab_1      | 1|ai     |   File "/appsrc/ai/recommender/surprise_recommenders.py", line 97, in update_training_data
lab_1      | 1|ai     |
lab_1      | 1|ai     | self.reader, rating_scale=(0,1))
lab_1      | 1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 193, in load_from_df
lab_1      | 1|ai     |
lab_1      | 1|ai     | return DatasetAutoFolds(reader=reader, df=df, rating_scale=rating_scale)
lab_1      | 1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 317, in __init__
lab_1      | 1|ai     |
lab_1      | 1|ai     | self.df.itertuples(index=False)]
lab_1      | 1|ai     |   File "/usr/local/lib/python3.7/site-packages/surprise/dataset.py", line 316, in <listcomp>
lab_1      | 1|ai     |
lab_1      | 1|ai     | for (uid, iid, r) in
lab_1      | 1|ai     | TypeError: float() argument must be a string or a number, not 'tuple'
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/roc_curve.json
machine_1  | 0|machine  | pushing file/appsrc/machine/learn/tmp/SVC/5df9468936a9990035cb1cfe/value.json
lab_1      | 0|lab    | serverSocket.emitEvent('expUpdated', '[object Object]')
```

**Comment by weixuanfu** [comment]
New dataset card UI for both regression and classification datasets 
![image](https://user-images.githubusercontent.com/21084970/72082644-78d38000-32ce-11ea-92e9-d297ea981573.png)

**Comment by weixuanfu** [comment]
I pushed a commit https://github.com/EpistasisLab/pennai/commit/bd0a5c409401b62d91e6afb245166fc8efae5168 for adding popup messages for those icons on MSE and MAE table. The commit message has a error on issue reference (it should be#229 instead of #129)

**Comment by hjwilli** [comment]
This is complete, and was released in 171285a62cc379663eda77adeb0a353317778316
